### PR TITLE
First draft of container support.

### PR DESCRIPTION
**Pull request by sdegler**
Schema (validator),  Schema(datomic), and input code needed to add container support to cook.
Still has spurious pprints in, these need to be removed prior to commit.
Clojure style and idiom comments more than welcome.

### CODE REVIEW

File: scheduler/src/cook/mesos/api.clj [hunk 1]
```diff
@@ -57,69 +83,167 @@
...
+               ])
+      :else
+      {})))
+
 (sm/defn submit-jobs
   [conn jobs :- [Job]]
-  (doseq [{:keys [uuid command max-retries max-runtime priority cpus mem user name ports uris env]} jobs
+  (doseq [{:keys [uuid command max-retries max-runtime priority cpus mem user name ports uris env container]} jobs
           :let [id (d/tempid :db.part/user)
-                ports (mapv (fn [port]
```
Comment (dgrnbrg) [inline scheduler/src/cook/mesos/api.clj]: Can you return this indent as it was so that I can see th e diff better? Thanks!

File: scheduler/src/cook/mesos/api.clj [hunk 2]
```diff
@@ -274,6 +440,21 @@
...
+;;; With a docker container ...
+;;; {"jobs": [{
+;;;  "command" : "/bin/echo $MYENV Hello World"
+;;;  "uuid"  : ""6e0eeb86-8c93-11e5-adfd-f8b156b232c9""
+;;;   "max-retries" : 3
+;;;   "uris" : [{"value" : "file:///home/sdegler/0.pref"}]
+;;;   "env" : "MYENV abc"
+;;;   "container" : { "type" : "docker",
+;;;                   "docker" : { "image" : "bp4_res_20150904.patched" ,
+;;;                                "parameters" : [ {"env" : "KRB5CCNAME=${KRB5CCNAME}" }, {"user" : "10138"}],
```
Comment (dgrnbrg) [inline scheduler/src/cook/mesos/api.clj]: why is there a second "env" in here?

File: scheduler/src/cook/mesos/util.clj [hunk 3]
```diff
@@ -28,6 +28,49 @@
...
+                       :else
+                       m))
+          fixup-keywords (fn fixup-keywords [m]
+                           (cond
+                             (map? m)
+                             (into {} (map (fn [p]
+                                             (let [k (first p)
+                                                   v (second p)]
+                                               [ (if (keyword? k)
+                                                   (keyword(name k)) k)
```
Comment (dgrnbrg) [inline scheduler/src/cook/mesos/util.clj]: we typically prefer to explicitly list the from and to keywords--although the dynamic casts are shorthand for now, they're harder to understand 6 months down the line.

File: scheduler/src/cook/mesos/api.clj (hunk 2 above)
Comment (sdegler) [inline scheduler/src/cook/mesos/api.clj]: its getting passed to the docker command line itself, the  example is from what I had been doing with marathon.
It is not necessarily needed, but would always be available.

File: scheduler/src/cook/mesos/util.clj (hunk 3 above)
Comment (sdegler) [inline scheduler/src/cook/mesos/util.clj]: I was thinking that maintaining the datomic/name -> name mapping would allow the code continue to work even as the schema gets extended under new versions of mesos.

Would you consider documentation as a solution or do you want  a keyword -> keyword table ?

Comment (dgrnbrg) [inline scheduler/src/cook/mesos/util.clj]: I'd prefer explicit destructuring and reconstruction, similar to how job->task-info is also very explicit.

**Comment by sdegler** [comment]
I could have cleaned up the destructuring using a table driven approach 
{ old-kw ( new-kw function-to-call-on-value )}
but I felt like this would get opaque as well in the same fashion as my original function.

**Comment by sdegler** [comment]
I bumped into clojure.walk/{prewalk,walk,postwalk} as well and was surprised that none actually really did exactly what I thought they would.

**Comment by dgrnbrg** [comment]
We try to explicitly write the code to do the format conversion. Explicitly pulling all the fields out of one entity and then populating specific fields in the new map has repeatedly helped us to avoid classes of bugs that automatic conversion admits.

**Comment by sdegler** [comment]
So are you happy with the newer version?  I'll remove the older commented one then.

**Comment by dgrnbrg** [comment]
I think that the newer version could be clarified by refactoring it into several functions, rather than one really big function.

**Comment by sdegler** [comment]
I ended up doing the table driven version, but I think its pretty clear whats going on.
I also broke it out.
This version is *not* segfaulting in mesos now. So give it a good look and please merge.

File: scheduler/src/cook/mesos/api.clj [hunk 4]
```diff
@@ -57,15 +83,98 @@
...
+
+;; helper for submit-jobs, deal with container structure
+(defn build-container [id container]
+  (let [container-id (d/tempid :db.part/user)
+        docker-id (d/tempid :db.part/user)
+        ctype (:type container)
+        volumes (if (contains? container :volumes)
+                  (:volumes container)
+                  [])
+        mkparams (fn [cid params]
```
Comment (dgrnbrg) [inline scheduler/src/cook/mesos/api.clj]: can you move these functions into top-levels?

File: scheduler/src/cook/mesos/api.clj [hunk 5]
```diff
@@ -57,15 +83,98 @@
...
+                                           ]
+                                        {:docker.param/key k
+                                         :docker.param/value v}
+                                        )) params)})))
+        mkvolumes (fn [cid vols]
+                    (let [modemap { "RO" 2 "RW" 1 }]
+                      (if (empty? vols) {}
+                          {:container/volumes
+                           (mapv
+                            (fn [m]
```
Comment (dgrnbrg) [inline scheduler/src/cook/mesos/api.clj]: Can you use map destructuring here instead? Like

```clojure
(fn [{:keys [container_path host_path mode]}]
    {:container.volume/container_path container_path
    ...
    })
```

File: scheduler/src/cook/mesos/api.clj [hunk 6]
```diff
@@ -57,15 +83,98 @@
...
+                                         ))) {} m))vols)})))]
+    (cond
+      (= (clojure.string/lower-case ctype) "docker")
+      (let [docker (:docker container)
+            params (if (contains? docker :parameters)
+                     (:parameters docker)
+                     [])
+            netmap { "HOST" 1 "BRIDGE" 2 "NONE" 3}]
+        [[:db/add id :job/container container-id]
+         (into {:db/id container-id
```
Comment (dgrnbrg) [inline scheduler/src/cook/mesos/api.clj]: this should be part of the `mkvolumes`

File: scheduler/src/cook/mesos/api.clj [hunk 7]
```diff
@@ -57,15 +83,98 @@
...
+                :container/type "DOCKER"}
+               ;;:container/type 1 }
+               (mkvolumes container-id volumes))
+         [:db/add container-id :container/docker docker-id]
+         (into {:db/id docker-id
```
Comment (dgrnbrg) [inline scheduler/src/cook/mesos/api.clj]: this should be part of the `mkparams` (which should probably have a different name)

File: scheduler/src/cook/mesos/api.clj [hunk 8]
```diff
@@ -73,9 +182,9 @@
                             ports)
                 uris (mapcat (fn [{:keys [value executable? cache? extract?]}]
                                (let [uri-id (d/tempid :db.part/user)
-                                     optional-params {:resource.uri/executable? executable?
+                                     optional-params [:resource.uri/executable? executable?
```
Comment (dgrnbrg) [inline scheduler/src/cook/mesos/api.clj]: why did you change this?

File: scheduler/src/cook/mesos/api.clj [hunk 9]
```diff
@@ -135,8 +250,38 @@
 (defn validate-and-munge-job
   "Takes the user and the parsed json from the job and returns proper
    Job objects, or else throws an exception"
-  [db user task-constraints {:strs [cpus mem uuid command priority max_retries max_runtime name uris ports env] :as job}]
-  (let [munged (merge
+  [db user task-constraints {:strs [cpus mem uuid command priority max_retries max_runtime name uris ports env container] :as job}]
+  (let [make-docker (fn [d]
+                      (merge { :image (get d "image") }
+                             (if (contains? d "network")
```
Comment (dgrnbrg) [inline scheduler/src/cook/mesos/api.clj]: use `if-let`

File: scheduler/src/cook/mesos/schema.clj [hunk 10]
```diff
@@ -72,36 +72,111 @@
...
+     :db/cardinality :db.cardinality/one
+     :db.install/_attribute :db.part/db}
+    {:db/id (d/tempid :db.part/db)
+     :db/ident :docker.param/value
+     :db/valueType :db.type/string
+     :db/cardinality :db.cardinality/one
+     :db.install/_attribute :db.part/db}
+    ;; Container Volume Attributes
+    {:db/id (d/tempid :db.part/db)
+     :db/ident :container.volume/container_path
```
Comment (dgrnbrg) [inline scheduler/src/cook/mesos/schema.clj]: use dashes, not underscores

File: scheduler/src/cook/mesos/util.clj [hunk 11]
```diff
@@ -28,6 +28,69 @@
...
+        ;; tables are { :datomic-kw :mesos-kw }
+        ;;             -- or --
+        ;;            { :datomic-kw [ mesos-kw fn-to-call-on-value ]}
+        ;; the functions are partially bound to the appropriate table to
+        ;; drive the parsing
+        volume-table {:container.volume/host_path :host_path
+                      :container.volume/container_path :container_path
+                      :container.volume/mode :mode}
+
+        param-table {:docker.param/key :key
```
Comment (dgrnbrg) [inline scheduler/src/cook/mesos/util.clj]: please use destructuring instead of table lookups

File: scheduler/src/cook/mesos/api.clj [hunk 12]
```diff
@@ -57,15 +83,98 @@
...
+        mkvolumes (fn [cid vols]
+                    (let [modemap { "RO" 2 "RW" 1 }]
+                      (if (empty? vols) {}
+                          {:container/volumes
+                           (mapv
+                            (fn [m]
+                              (reduce
+                               (fn [b kvp]
+                                 (into b
+                                       (cond
```
Comment (dgrnbrg) [inline scheduler/src/cook/mesos/api.clj]:
```clojure
(into {}
(when foo {foo foo_value}))
```

File: scheduler/project.clj [hunk 13]
```diff
@@ -86,7 +86,7 @@
                  [liberator "0.13"]
 
                  ;;Databases
-                 [com.datomic/datomic-free "0.9.5206"
+                 [com.datomic/datomic-pro "0.9.5153"
```
Comment (icexelloss) [inline scheduler/project.clj]: Revert datomic version

File: scheduler/project.clj [hunk 14]
```diff
@@ -16,7 +16,7 @@
 (defproject cook "1.0.1-SNAPSHOT"
   :description "This launches jobs on a Mesos cluster with fair sharing and preemption"
   :license {:name "Apache License, Version 2.0"}
-  :dependencies [[org.clojure/clojure "1.6.0"]
+  :dependencies [[org.clojure/clojure "1.7.0"]
```
Comment (icexelloss) [inline scheduler/project.clj]: Can you try with 1.6?

File: scheduler/project.clj [hunk 15]
```diff
@@ -94,6 +94,7 @@
                                org.slf4j/log4j-over-slf4j
                                org.slf4j/slf4j-nop
                                joda-time]]
+                 [com.basho.riak/riak-client "1.4.4"]
```
Comment (icexelloss) [inline scheduler/project.clj]: Do we need this?

File: scheduler/src/cook/mesos/api.clj [hunk 16]
```diff
@@ -57,69 +83,187 @@
...
+;;   :uris [{:value "file:///home/sdegler/0.pref"}]
+;;   :env "MYENV abc"
+;;   :container { :type "docker"
+;;     :docker {:image "bp4_res_20150904.patched"
+;;              :parameters [ {:env "KRB5CCNAME=${KRB5CCNAME}" } {:user "10138"}]
+;;              :network "HOST"
+;;             }
+;;   :volumes [ {:containerPath "/tmp/$KRB5CCNAME} :hostPath "/tmp/$KRB5CCNAME" :mode "RO"}]
+;;
+;; }
```
Comment (icexelloss) [inline scheduler/src/cook/mesos/api.clj]: Make this as a clojure comment block

File: scheduler/src/cook/mesos/api.clj [hunk 17]
```diff
@@ -57,69 +83,187 @@
...
+;;
+;;
+;;
+
+(defn mk-container-params  [cid params]
```
Comment (icexelloss) [inline scheduler/src/cook/mesos/api.clj]: [cid params] should be at next line

File: scheduler/src/cook/mesos/api.clj [hunk 18]
```diff
@@ -57,69 +83,187 @@
...
-                     :job/command command
-                     :job/custom-executor false
-                     :job/user user
-                     :job/max-retries max-retries
-                     :job/state :job.state/waiting
-                     :job/resource [{:resource/type :resource.type/cpus
-                                     :resource/amount cpus}
-                                    {:resource/type :resource.type/mem
-                                     :resource/amount mem}]}]]
+          ports (mapv (fn [port]
```
Comment (icexelloss) [inline scheduler/src/cook/mesos/api.clj]: indentation is off

File: scheduler/src/cook/mesos/api.clj (hunk 16 above)
Comment (icexelloss) [inline scheduler/src/cook/mesos/api.clj]: put:

(build-container json) in a comment block

File: scheduler/src/cook/mesos/api.clj [hunk 19]
```diff
@@ -57,69 +83,187 @@
...
+                      :job/custom-executor false
+                      :job/user user
+                      :job/max-retries max-retries
+                      :job/state :job.state/waiting
+                      :job/resource [{:resource/type :resource.type/cpus
+                                      :resource/amount cpus}
+                                     {:resource/type :resource.type/mem
+                                      :resource/amount mem}]}]]]
+                                        ;debug
+    (log/info (clojure.pprint/write (-> ports
```
Comment (icexelloss) [inline scheduler/src/cook/mesos/api.clj]: Remove this

File: scheduler/src/cook/mesos/api.clj [hunk 20]
```diff
@@ -135,8 +279,55 @@
 (defn validate-and-munge-job
   "Takes the user and the parsed json from the job and returns proper
    Job objects, or else throws an exception"
-  [db user task-constraints {:strs [cpus mem uuid command priority max_retries max_runtime name uris ports env] :as job}]
-  (let [munged (merge
+  [db user task-constraints {:strs [cpus mem uuid command priority max_retries max_runtime name uris ports env container] :as job}]
+  (let [make-docker
```
Comment (icexelloss) [inline scheduler/src/cook/mesos/api.clj]: Let's try if http://stackoverflow.com/questions/9406156/clojure-convert-hash-maps-key-strings-to-keywords can achieve the same thign here

File: scheduler/src/cook/mesos/api.clj [hunk 21]
```diff
@@ -147,10 +338,22 @@
...
+                                        {})
+                                      (if (contains? container "volumes")
+                                        {:volumes (make-volumes
+                                                   (get container "volumes"))}
+                                        {}))})
                  (when uris
                    {:uris (map (fn [{:strs [value executable cache extract]}]
                                  (merge {:value value}
-                                        (when executable {:executable? executable})
+                                        (when executable
```
Comment (icexelloss) [inline scheduler/src/cook/mesos/api.clj]: make it the same line

File: scheduler/src/cook/mesos/api.clj [hunk 22]
```diff
@@ -147,10 +338,22 @@
                   :ports (or ports [])
                   :cpus (double cpus)
                   :mem (double mem)}
+
+                 (when container
+                   {:container (merge { :type (get container "type")}
```
Comment (icexelloss) [inline scheduler/src/cook/mesos/api.clj]: try keywordize-keys

File: scheduler/src/cook/mesos/api.clj [hunk 23]
```diff
@@ -274,6 +477,21 @@
 ;;;            "cpus": 1.5,
 ;;;            "mem": 1000}]}
 ;;;
+;;; With a docker container ...
```
Comment (icexelloss) [inline scheduler/src/cook/mesos/api.clj]: Please use a comemnt block

File: scheduler/src/cook/mesos/scheduler.clj [hunk 24]
```diff
@@ -83,23 +84,30 @@
...
-     :task-id task-id
-     :num-ports (count (:ports resources))
-     :resources (select-keys resources [:mem :cpus])
-     ;;TODO this data is a race-condition
-     :data (.getBytes
-             (pr-str
-               {:instance (str (count (:job/instance job-ent)))})
-             "UTF-8")
-     executor-key executor-value}))
+                         (into {:executor-id (str (java.util.UUID/randomUUID))
```
Comment (icexelloss) [inline scheduler/src/cook/mesos/scheduler.clj]: try 
(-> {...}
      #(when container (assoc % :container container)))

File: scheduler/src/cook/mesos/scheduler.clj [hunk 25]
```diff
@@ -83,23 +84,30 @@
...
+                                :framework-id fid
+                                :name "cook agent executor"
+                                :source "cook_scheduler"
+                                :command command}
+                               (if (seq container)
+                                 {:container container}
+                                 {}))
+                         command)
+        ;; If the there is no value for key :job/name, the following name will contain a substring "null".
+        task (into {:name (format "%s_%s_%s" (:job/name job-ent "cookjob") (:job/user job-ent) task-id)
```
Comment (icexelloss) [inline scheduler/src/cook/mesos/scheduler.clj]: probably same as above

File: scheduler/src/cook/mesos/scheduler.clj [hunk 26]
```diff
@@ -190,8 +197,8 @@
                                      :task-killed
                                      :task-lost} :instance.status/failed)
                  prior-job-state (:job/state (d/entity db job))
-                 progress (when (:data status)
-                            (:percent (read-string (String. (:data status)))))
+                 progress nil #_(when (:data status)
```
Comment (icexelloss) [inline scheduler/src/cook/mesos/scheduler.clj]: revent this

File: scheduler/src/cook/mesos/scheduler.clj [hunk 27]
```diff
@@ -73,6 +73,7 @@
         resources (util/job-ent->resources job-ent)
         ;; If the custom-executor attr isn't set, we default to using a custom
         ;; executor in order to support jobs submitted before we added this field
+        container (util/job-ent->container db job job-ent)
```
Comment (icexelloss) [inline scheduler/src/cook/mesos/scheduler.clj]: please change to (container-ent->container db (:job/container job-ent)))

File: scheduler/src/cook/mesos/util.clj [hunk 28]
```diff
@@ -28,6 +28,98 @@
...
+                       (let [sm (filter (fn [p]
+                                          (not(= :db/id (first p)))) m)]
+                         (into {} (map (fn [p]
+                                         [(first p) (rm-dbids (second p))])
+                                       sm)))
+                       (vector? m)
+                       (mapv rm-dbids m)
+                       :else
+                       m))
+          cmap (d/pull db "[*]" ceid)]
```
Comment (icexelloss) [inline scheduler/src/cook/mesos/util.clj]: Pass a pattern in here

**Comment by wyegelwel** [comment]
Degler, I'm closing this PR in favor of #135 which was based off of this code. If you get a chance, please review and comment on the new pull request.
